src/Compilers/CSharp/Portable/Binder/Binder_Lambda.cs

```diff
@@ -422,5 +423,61 @@ private UnboundLambda BindAnonymousFunction(AnonymousFunctionExpressionSyntax sy
 
             return lambda;
         }
+
+        // Please don't use thread local storage widely. This should be one of only a few uses.
+        [ThreadStatic] private static PooledDictionary<SyntaxNode, int>? s_lambdaBindings;
+
+        internal TResult BindWithLambdaBindingCountDiagnostics<TSyntax, TArg, TResult>(
+            TSyntax syntax,
+            TArg arg,
+            BindingDiagnosticBag diagnostics,
+            Func<Binder, TSyntax, TArg, BindingDiagnosticBag, TResult> bind)
+            where TSyntax : SyntaxNode
+            where TResult : BoundNode
+        {
+            Debug.Assert(s_lambdaBindings is null);
+            var bindings = PooledDictionary<SyntaxNode, int>.GetInstance();
+            s_lambdaBindings = bindings;
+
+            try
+            {
+                TResult result = bind(this, syntax, arg, diagnostics);
+
+                foreach (var pair in bindings)
+                {
+                    const int maxLambdaBinding = 100;
```

Comment:
This is essentially a magic number. We should really have a comment here discussing why this magic number was chosen and what others should consider when changing it in the future.

Reply:
Added comment in https://github.com/dotnet/roslyn/pull/73050.